src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Cipher.java

```diff
@@ -988,6 +1119,79 @@ private int implDoFinal(ByteBuffer outBuffer)
         }
     }
 
+    private int getBytesToCompleteBlock(int availableBytes) {
+        int partBlock = availableBytes & (blockSize - 1);
+        return partBlock == 0 ? 0 : blockSize - partBlock;
+    }
+
+    private int getCTSMustBeBuffered(int availableBytes) {
+        return Math.min(availableBytes,
+                padBuffer.length - getBytesToCompleteBlock(availableBytes));
+    }
+
+    /**
+     * The ciphertext ordering for the three variants can be depicted as
+     * follows, where 'p' is the penultimate block (which may be partial
+     * or full), and 'f' the full final block:
+     *
+     *                    'p' is a partial block   'p' is a full block
+     *                   ------------------------ ---------------------
+     *   CS1 (NIST)     |     .... pp ffff       |    .... pppp ffff
+     *   CS2 (Schneier) |     .... ffff pp       |    .... pppp ffff
+     *   CS3 (Kerberos) |     .... ffff pp       |    .... ffff pppp
+     *
+     * After encryption, we get the ciphertext from the token formatted as
+     * specified in the SunPKCS11 'cipherTextStealingVariant' configuration
+     * property. Conversely, before decryption, the ciphertext has to be passed
+     * to the token according to the previous formatting. This method converts
+     * the ciphertext between the format used by the token and the one used by
+     * SunJCE's "AES/CTS/NoPadding" implementation (CS3 as described by RFC
+     * 2040, section 8).
+     */
+    private void convertCTSVariant(ByteBuffer ciphertextBuf,
+            byte[] ciphertextArr, int end) {
+        if (end == blockSize) {
+            // No reordering needed for a single block
+            return;
+        }
+        assert token.ctsVariant != null : "CTS algorithms should not be " +
+                "registered if the CTS variant of the token is unknown";
+        if (token.ctsVariant == Token.CTSVariant.CS3) {
+            // Already CS3
+            return;
+        }
+        int pad = end % blockSize;
+        if (token.ctsVariant == Token.CTSVariant.CS2 && pad != 0) {
+            // CS2 and 'p' is a partial block, equal to CS3
+            return;
+        }
+        if (ciphertextArr != null) {
+            ciphertextBuf = ByteBuffer.wrap(ciphertextArr);
+        }
+        if (ciphertextBuf != null) {
+            pad = pad == 0 ? blockSize : pad;
+            byte[] tmp = new byte[blockSize];
+            int start = end - blockSize - pad;
+            if (encrypt) {
+                // .... pp[pp] ffff -> .... ffff pp[pp]
+                // Ultimate ffff block to the temporary buffer
+                ciphertextBuf.get(end - blockSize, tmp);
+                // Penultimate pp[pp] block to the end of ciphertext
+                ciphertextBuf.put(end - pad, ciphertextBuf, start, pad);
+                // Temporary buffer to the penultimate block
+                ciphertextBuf.put(start, tmp);
+            } else {
```

Comment:
Personally, I find it easier to follow if this code block follows the decrypt case (line 1184-1190), the allocated `tmp` could be smaller, e.g.
```suggestion
                byte[] tmp = new byte[pad];
                // .... pp[pp] ffff -> .... ffff pp[pp]
                ciphertextBuf.get(start, tmp);
                ciphertextBuf.put(start, ciphertextBuf, end - blockSize, blockSize);
                ciphertextBuf.put(end - pad, tmp);
```

Reply:
I have no personal preference, but would suggest that if we change it to cut the pad, we keep the decryption case aligned.

Reply:
Looks good to me.

Reply:
Yes, I prefer this.

Reply:
Done in 5fe83b20a10002c4c48e90321e69629bbd0aa9ff.